README.md
PlayMaker--Utils
================

## Licensing
This package is released under LGPL license: [http://opensource.org/licenses/LGPL-3.0](http://opensource.org/licenses/LGPL-3.0)


## Description
This is the authoritative source code for "PlayMaker Utils", used in many different projects, like ArrayMaker, XmlMaker, Mecanim, Pathfinding, UI and other Unity modules.

## Implementation
More infos on how to use this repository in your own project using Git and SubModules. 
Refer to the 2dPlatformer UnityLearn PlayMaker project to see how it's integrated. 
Instead of making hard copy, I use this repository as a sub module, allowing me to work on this utils, add functions and improve it without having to switch projects or manually integrate changes from one project to another.

## Including PlayMaker Utils in your Github Rep as a Sub Module using SourceTree
1: Delete the existing "PlayMaker Utils" folder  
2: commit this deletion  
3: in the side bar of SourceTree, right click "Add SubModule"  
4: SourcePath/Url should be https://github.com/jeanfabre/PlayMaker--Utils.git 
5: set the Locale Relative Path to "Assets/PlayMaker Utils"  
6: Click Ok, let sourcetree import this submodule  

Now, make sure you "show" the submodules when you work on that project, and make visible this submodule entry in the side bar, it will prompt you with a pull icon when needed.

## Pull a SubModule
Simple double click on the submodule entry in the side bar, it will open a SourceTree window of that rep, and pull as usual. Close this window when you are done pulling.

## pushing to this Submodule
Simple Double click on the submodule entry in the side bar, it will open a SourceTree window of that rep, and commit/push as usual. Contact me to get writing access and be allow to push.



Changelog.md
#PlayMaker Utils Change log

###1.6.3
**Release Date:** 12/04/2021

**fixed**: fixed PlayMakerEvent and PlayMakerTimelineEventTarget to support all new versions of unity starting from 2017


###1.6.2
**Release Date:** 04/10/2020  

**fixed**: fixed playmaker 1.9 support for enums ( used in `GetEventProperties` for example)


###1.6.1
**Release Date:** 04/10/2019  

**New**: New method `SendEventToGameObjectFsmByName(...)` to send an event to a specific fsm by name

**fixed**: fixed Enum Creator Wizard editing refresh and ordered of enum list.

###1.6.0
**Release Date:** 26/07/2019  

**New**: New method `PlayMakerEditorUtils.CreateGlobalEventsIfNeeded(...)` to create global asset, stored in PlayMaker Globals asset

###1.5.9
**Release Date:** 05/04/2019  

**Update**: switched from LGPL to MIT license

###1.5.8
**Release Date:** 17/01/2019  

**fixed**: Updated Enum Creator Wizard to work on Unity 2018

###1.5.7
**Release Date:** 18/12/2018  

**fixed**: Removed Obsolete EventType.mouseDown api usage


###1.5.6
**Release Date:** 18/09/2018  

**fixed**: Brought back `GetFsmOnGameObject` method to compliment `FindFsmOnGameObject` and return an Fsm instead of PlayMakerFSM object, this is to ease updates of old projects.

###1.5.5
**Release Date:** 30/08/2018  

**new**: `PlayMakerUtils.LogFullPathToAction(FsmStateAction)`  
**new**: `PlayMakerUtils.LogFullPathToAction.GetGameObjectPath(GameObject)`  


###1.5.4
**Release Date:** 20/07/2018  

**new**: VersionInfo class to help finding out addons versions and misc

###1.5.3
**Release Date:** 18/07/2018  

**fixed**: removed unwanted Editor reference that prevented builds

**fixed**: leveraged global event creation with better checks


###1.5.2
**Release Date:** 27/06/2018  

**new**: new `Current Event Data` for viewing live the current event data being used.

###1.5.1
**Release Date:** 11/06/2018  

**Improvement** `EventProxyWizard` now can accepts a single parameter passed as `EventData` to the playmaker event


###1.5
**Release Date:** 20/03/2018  

**new**: new `PlayMakerTimelineEventTarget` for creating convenient proxies for timeline

**new**: debug optional parameter for `PlayMakerEvent` `sendevent()` call

**new**: Editor Utils to copy string to clipboard `PlayMakerEditorUtils.CopyTextToClipboard()`

**Fix**: Fixed mounting and unmounting scripting symbols, properly based on PlaymakerDefines.cs. 

**Fix**: ProceduralMaterial support rule for 2017 and 2018

**new**: `PLAYMAKER_UTILS_1_5_OR_NEWER` symbol

###1.4.2
**Release Date:** 22/06/2017  

**new**: new `PlayMakerUtils_conversions` to convert non nullable objects like rect, quaternions, vectors, from an object pointer.  

###1.4.1
**Release Date:** 05/06/2017  

**new**: new `PlayMakerUtils.SendEventToTarget` to leverage sending events


###1.4
**Release Date:** 01/06/2017

**new**: PlayMakerUtilsDefine class to feature `PLAYMAKER_UTILS` and `PLAYMAKER_UTILS_1_4_OR_NEWER`  

**new**: MainCamerTarget Public Class with property drawer like the owner class for proxy component creation  

**Fix**: owner public class and property drawer to work as expected

**Fix**: `SetEventProperties.Properties` is now declared by default to always be there  

###1.3.2
**Release Date:** 07/03/2017

**Fix**: Windows Publishing with a better code to check if type is enum  

**new**: Improved `DoesGameObjectImplementsEvent()` to account for `SendToChildren` Option

###1.3.1
**Release Date:** 06/02/2017

**Fix**: `Required` Attribute definition now properly namespaced in `HutongGames.PlayMaker.Ecosystem.Utils`

**Fix**: Enum Creator wizard title  

###1.3
**Release Date:** 12/01/2017

**new**: Enum Creator Wizard  
**new**: Rotorz ReOrderable List Library included 

###1.2.9
**Release Date:** 27/09/2016  

**new**: Conditional Expression addition ( manual for now)  

**Fix**: Windows store publishing  

###1.2.8
**Release Date:** 21/09/2016  

**new**: RequiredAttribute  

**update**: LinkerWizard support for PlayMaker 1.8.3  

**Improvement:** LinkerWizard do not spam logs when published, debug all actions set to false by default 


###1.2.7
**Release Date:** 30/08/2016  

**new**: added `TransformEventsBridge.cs` to forward `OnTransformParentChanged()` and `OnTransformChildrenChanged()` as PlayMaker Events.

###1.2.6
**Release Date:** 17/08/2016  

**new**: added PlayMakerInspectorUtils.SetActionEditorVariableSelectionContext() to properly set context when using VariableEditor.FsmXXXField(). This is for PlayMaker 1.8+ only  

###1.2.5
**Release Date:** 22/06/2016  

**Fix**: PlayMaker 1.8 support

###1.2.4 beta
**Release Date:** 23/03/2016  

**Fix**: Prevent error when parsing a null object as string in ParseFsmVarToString()  

###1.2.3 beta
**Release Date:** 01/02/2016

**Fix**: Linker wizard support for Collider2D inheritance.

###1.2.2 beta
**Release Date:** 12/01/2016  

**Fix:** Unity 4.7 compatibility 

###1.2.1 beta
**Release Date:** 14/12/2015  

**Fix:** Unity 5.3 new SceneManagement obsolete calls  
**Fix:** PlayMaker 1.8f36 change in EventTarget setup for sending events programmatically

###1.2.0 beta
**Release Date:** 11/12/2015  

**New:** New Component public class with related propertyDrawers for `PlayMakerFsmVariableTarget` and `PlayMakerFsmVariable`  
**New:** New utils to list variables by string based on a FsmVariables reference   
**New:**  New Reflections Utils to get the BaseProperty of a SerializedProperty   

###1.1.6
**Release Date:** 19/09/2015  

**New:** New function to create global events  
**Fix:** Added support for PlayMakerEvent propertyDrawer to work within stateMachine context  
**Improvement:** Better description for auto generated "PlayMaker sent even proxy"  
  
###1.1.5
**Release Date:**  16/09/2015  

**New:** Event Proxy Creator wizard  

###1.1.4
**Release Date:**  10/09/2015  

**New:** Custom Sample for linker wizard  
**Improvement:** Moved LinkerWizard package into U4 github rep to centralize content  

Editor/PlayMakerEditorUtils.cs
// (c) Copyright HutongGames, LLC 2010-2015. All rights reserved.

using UnityEngine;
using UnityEditor;

using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;

using HutongGames;
using HutongGames.PlayMaker;
using HutongGames.PlayMakerEditor;


public class PlayMakerEditorUtils : Editor {
	
	#if UNITY_4_6 || UNITY_5_0 || UNITY_5_1 || UNITY_5_2
	[MenuItem ("PlayMaker/Addons/Tools/Export Current Scene",false,100)]
	public static void ExportCurrentScene()
	{

		if (!EditorApplication.SaveCurrentSceneIfUserWantsTo())
		{
			return;
		}
		
		EditorUtility.DisplayDialog("PlayMaker dll export","Just a reminder that PlayMaker.dll file is not redistributable,\n\nMake sure you uncheck: \n'Assets/PlayMaker/PlayMaker.dll'\n\nwhen exporting a package to sharing with others.","Ok");
		
		
		var _sel =  Selection.objects;
		
		
		if (EditorUtility.DisplayDialog("Export Globals?","If your scene if using global variables, it will need to be included in the package as well.","Export Globals","Ignore Globals"))
		{
			EditorApplication.ExecuteMenuItem("PlayMaker/Tools/Export Globals");
			var _globalAsset = AssetDatabase.LoadAssetAtPath("Assets/PlaymakerGlobals_EXPORTED.asset",typeof(UnityEngine.Object));
			ArrayUtility.Add<UnityEngine.Object>(ref _sel ,_globalAsset);
			Selection.objects = _sel;
		}
		
	
		SelectSceneCustomAction();
		
		
	
		var _scene = AssetDatabase.LoadAssetAtPath(EditorApplication.currentScene,typeof(UnityEngine.Object));
		_sel =  Selection.objects;
		ArrayUtility.Add<UnityEngine.Object>(ref _sel ,_scene);
		Selection.objects = _sel;
		EditorApplication.ExecuteMenuItem("Assets/Export Package...");
	}
	
	
	[MenuItem ("PlayMaker/Addons/Tools/Export Current Scene", true)]
	public static bool CheckExportCurrentScene() {
	    return !String.IsNullOrEmpty(EditorApplication.currentScene);
	}
	
	
	[MenuItem ("PlayMaker/Addons/Tools/Select Current Scene Used Custom Actions")]
	public static void SelectSceneCustomAction()
	{
		UnityEngine.Object[] _list = GetSceneCustomActionDependencies();
		
		var _sel =  Selection.objects;
		ArrayUtility.AddRange<UnityEngine.Object>(ref _sel ,_list);
		Selection.objects = _sel;	
	}

	#endif

	public static UnityEngine.Object[] GetSceneCustomActionDependencies()
	{
		
		UnityEngine.Object[] list = new UnityEngine.Object[0];
		
		FsmEditor.RebuildFsmList();

		List<PlayMakerFSM> fsms = FsmEditor.FsmComponentList;
		
//		List<System.Type> PlayMakerActions = FsmEditorUtility.Actionslist;

		foreach(PlayMakerFSM fsm in fsms)
		{
			
			//Debug.Log(FsmEditorUtility.GetFullFsmLabel(fsm));
			
			//if (fsm.FsmStates != null) fsm.FsmStates.Initialize();
			
			for (int s = 0; s<fsm.FsmStates.Length; ++s)
			{
				
					fsm.FsmStates[s].LoadActions();
				
					Debug.Log(fsm.FsmStates[s].Name+" is loaded:"+fsm.FsmStates[s].ActionsLoaded);
				
					// Show SendEvent and SendMessage as we find them
					foreach(FsmStateAction action in fsm.FsmStates[s].Actions)
					{
						#if PLAYMAKER_1_8_OR_NEWER
							UnityEngine.Object _asset = ActionScripts.GetAsset(action);
						#else
							UnityEngine.Object _asset = FsmEditorUtility.GetActionScriptAsset(action);
						#endif
						
						string _name = action.Name;
						if (String.IsNullOrEmpty(_name))
						{
							if (_asset!=null)
							{
								_name = _asset.name;
							}else
							{

								#if PLAYMAKER_1_8_OR_NEWER
									_name = Labels.GetActionLabel(action) + "[WARNING: FILE NOT FOUND]";
								#else
									_name = FsmEditorUtility.GetActionLabel(action) + "[WARNING: FILE NOT FOUND]";
								#endif

							}
							
						}
					
						if (Enum.IsDefined(typeof(WikiPages),_name))
						{
						//	Debug.Log(_name+" : official action");
						}else{
						//	Debug.Log(_name+" : custom action");
						
							if (_asset!=null)
							{
								ArrayUtility.Add<UnityEngine.Object>(ref list ,_asset);
							}
						}
							
					}
			}
			
			
		}
		
		return list;
		
	}// GetSceneCustomActionDependencies
	
	/// <summary>
	//	This makes it easy to create, name and place unique new ScriptableObject asset files.
	/// </summary>
	public static void CreateAsset<T> (string name="") where T : ScriptableObject
	{
		T asset = ScriptableObject.CreateInstance<T> ();
		
		string path = AssetDatabase.GetAssetPath (Selection.activeObject);
		if (path == "") 
		{
			path = "Assets";
		} 
		else if (Path.GetExtension (path) != "") 
		{
			path = path.Replace (Path.GetFileName (AssetDatabase.GetAssetPath (Selection.activeObject)), "");
		}

		string _name = string.IsNullOrEmpty(name)? "New " + typeof(T).ToString() : name ;

		string assetPathAndName = AssetDatabase.GenerateUniqueAssetPath (path + "/" + _name + ".asset");
		
		AssetDatabase.CreateAsset (asset, assetPathAndName);
		
		AssetDatabase.SaveAssets ();
		AssetDatabase.Refresh();
		EditorUtility.FocusProjectWindow ();
		Selection.activeObject = asset;
	}


	
	/// <summary>
	/// Used to get assets of a certain type and file extension from entire project
	/// </summary>
	/// <param name="type">The type to retrieve. eg typeof(GameObject).</param>
	/// <param name="fileExtension">The file extention the type uses eg ".prefab".</param>
	/// <returns>An Object array of assets.</returns>
	public static UnityEngine.Object[] GetAssetsOfType(System.Type type, string fileExtension)
	{
		List<UnityEngine.Object> tempObjects = new List<UnityEngine.Object>();
		DirectoryInfo directory = new DirectoryInfo(Application.dataPath);
		FileInfo[] goFileInfo = directory.GetFiles("*" + fileExtension, SearchOption.AllDirectories);
		
		int i = 0; int goFileInfoLength = goFileInfo.Length;
		FileInfo tempGoFileInfo; string tempFilePath;
		UnityEngine.Object tempGO;
		for (; i < goFileInfoLength; i++)
		{
			tempGoFileInfo = goFileInfo[i];
			if (tempGoFileInfo == null)
				continue;
			
			tempFilePath = tempGoFileInfo.FullName;
			tempFilePath = tempFilePath.Replace(@"\", "/").Replace(Application.dataPath, "Assets");
			
			//Debug.Log(tempFilePath + "\n" + Application.dataPath);
			
			tempGO = AssetDatabase.LoadAssetAtPath(tempFilePath, typeof(UnityEngine.Object)) as UnityEngine.Object;
			if (tempGO == null)
			{
				//Debug.LogWarning("Skipping Null");
				continue;
			}
			else if (tempGO.GetType() != type)
			{
				//Debug.LogWarning("Skipping " + tempGO.GetType().ToString());
				continue;
			}
			
			tempObjects.Add(tempGO);
		}
		
		
		
		
		return tempObjects.ToArray();
	}
	
	public static UnityEngine.Object GetAssetByName(string fileName)
	{
		DirectoryInfo directory = new DirectoryInfo(Application.dataPath);
		FileInfo[] goFileInfo = directory.GetFiles("*" + fileName, SearchOption.AllDirectories);
		
		int i = 0; int goFileInfoLength = goFileInfo.Length;
		FileInfo tempGoFileInfo; string tempFilePath;
		UnityEngine.Object tempGO;
		for (; i < goFileInfoLength; i++)
		{
			tempGoFileInfo = goFileInfo[i];
			if (tempGoFileInfo == null)
				continue;
			
			tempFilePath = tempGoFileInfo.FullName;
			tempFilePath = tempFilePath.Replace(@"\", "/").Replace(Application.dataPath, "Assets");
			
			//	Debug.Log(tempFilePath + "\n" + Application.dataPath);
			
			tempGO = AssetDatabase.LoadAssetAtPath(tempFilePath, typeof(UnityEngine.Object)) as UnityEngine.Object;
			if (tempGO == null)
			{
				Debug.LogWarning("Skipping Null");
				continue;
			}
			
			return tempGO;
		}
		
		return null;
	}


	/// <summary>
	/// Mounts the scripting define symbol to all targets. Taken from Hutonggames PlayMakerDefines.cs
	/// </summary>
	/// <param name="defineSymbol">Define symbol.</param>
	public static void MountScriptingDefineSymbolToAllTargets(string defineSymbol)
	{
		foreach (BuildTargetGroup group in Enum.GetValues(typeof(BuildTargetGroup)))
		{
			if (!IsValidBuildTargetGroup(group)) continue;
			
			var defineSymbols = PlayerSettings.GetScriptingDefineSymbolsForGroup(group).Split(';').Select(d => d.Trim()).ToList();
			if (!defineSymbols.Contains(defineSymbol))
			{
				defineSymbols.Add(defineSymbol);
				try
				{
					PlayerSettings.SetScriptingDefineSymbolsForGroup(group, string.Join(";", defineSymbols.ToArray()));
				}
				catch (Exception)
				{
					Debug.Log("Could not set PLAYMAKER defines for build target group: " + group);
					throw;
				}
				
			}
		}
	}

	/// <summary>
	/// Unmount scripting define symbol to all targets. Taken from Hutonggames PlayMakerDefines.cs
	/// </summary>
	/// <param name="defineSymbol">Define symbol.</param>
	public static void UnMountScriptingDefineSymbolToAllTargets(string defineSymbol)
	{
		foreach (BuildTargetGroup group in Enum.GetValues(typeof(BuildTargetGroup)))
		{
			if (!IsValidBuildTargetGroup(group)) continue;
			
			var defineSymbols = PlayerSettings.GetScriptingDefineSymbolsForGroup(group).Split(';').Select(d => d.Trim()).ToList();
			if (defineSymbols.Contains(defineSymbol))
			{
				defineSymbols.Remove(defineSymbol);
				PlayerSettings.SetScriptingDefineSymbolsForGroup(group, string.Join(";", defineSymbols.ToArray()));
			}
		}
	}

	private static bool IsValidBuildTargetGroup(BuildTargetGroup group)
	{
		if (group == BuildTargetGroup.Unknown || IsObsolete(group)) return false;
		
		// Checking Obsolete attribute should be enough, 
		// but sometimes Unity versions are missing attributes
		// so keeping these checks around just in case:
		
		#if UNITY_5_3_0 // Unity 5.3.0 had tvOS in enum but throws error if used
		if ((int)(object)group == 25) return false;
		#endif
		
		#if UNITY_5_4 || UNITY_5_5 // Unity 5.4+ doesn't like Wp8 and Blackberry any more
		if ((int)(object)group == 15) return false;
		if ((int)(object)group == 16) return false;
		#endif
		
		#if UNITY_5_6 // Unity 5.6+ removed build target 27?
		if ((int)(object)group == 27) return false;
		#endif
		
		return true;
	}
	
	private static bool IsObsolete(Enum value)
	{
		var enumInt = (int)(object)value;
		if (enumInt == 4 || enumInt == 14) return false;
		
		var field = value.GetType().GetField(value.ToString());
		var attributes = (ObsoleteAttribute[])field.GetCustomAttributes(typeof(ObsoleteAttribute), false);
		return attributes.Length > 0;
	}
	
	/* NOTE: IsObsolete is complicated by the definition of BuildTargetGroup enum. 
         * E.g., in Unity 5.4:
         * 
          public enum BuildTargetGroup
          {
            Unknown = 0,
            Standalone = 1,
            [Obsolete("WebPlayer was removed in 5.4, consider using WebGL")] WebPlayer = 2,
            iOS = 4,
            [Obsolete("Use iOS instead (UnityUpgradable) -> iOS", true)] iPhone = 4,
            PS3 = 5,
            XBOX360 = 6,
            Android = 7,
            WebGL = 13,
            [Obsolete("Use WSA instead")] Metro = 14,
            WSA = 14,
            [Obsolete("Use WSA instead")] WP8 = 15,
            [Obsolete("BlackBerry has been removed as of 5.4")] BlackBerry = 16,
            Tizen = 17,
            PSP2 = 18,
            PS4 = 19,
            PSM = 20,
            XboxOne = 21,
            SamsungTV = 22,
            Nintendo3DS = 23,
            WiiU = 24,
            tvOS = 25,
          }
  
         */


	public static void CopyTextToClipboard(string content)
	{
		TextEditor te = new TextEditor();
		te.content = new GUIContent(content);
		te.SelectAll();
		te.Copy();
	}

}
